osd/scheduler: Classify subOp reads according to op priority for mClock

sseshasa opened the pull request:
The change brings MSG_OSD_EC_READ into the fold of mClock scheduler. This
improves the scheduling of client and other classes of operation as they
are no longer unnecessarily preempted by the 'immediate' queue.
EC SubOps are now handled as follows:
    
 - EC SubOp reads generated during recovery will either go into the
   'background_recovery' or 'background_best_effort' class based on
   the recovery priority set for the op. EC SubOp reads generated due
   to client will continue to be classified as 'immediate'.

 - EC SubOp writes generated as a result of client operations will
   continue to be classified as 'immediate'.

 - EC SubOp replies are considered high priority and therefore
   continue to be classed as 'immediate'.

Fixes: https://tracker.ceph.com/issues/71655
Signed-off-by: Sridhar Seshasayee <sridhar.seshasayee@ibm.com>

<!--
  - Please give your pull request a title like

      [component]: [short description]

  - Please use this format for each git commit message:

      [component]: [short description]

      [A longer multiline description]

      Fixes: [ticket URL on tracker.ceph.com, create one if necessary]
      Signed-off-by: [Your Name] <[your email]>

    For examples, use "git log".
-->

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [x] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins test classic perf` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-perf-classic/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-perf-pull-requests/config/definitions/ceph-perf-pull-requests.yml)
- `jenkins test crimson perf` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-perf-crimson/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-perf-pull-requests/config/definitions/ceph-perf-pull-requests.yml)
- `jenkins test signed` [Jenkins Job](https://jenkins.ceph.com/job/ceph-pr-commits/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pr-commits/config/definitions/ceph-pr-commits.yml)
- `jenkins test make check` [Jenkins Job](https://jenkins.ceph.com/job/ceph-pull-requests/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pull-requests/config/definitions/ceph-pull-requests.yml)
- `jenkins test make check arm64` [Jenkins Job](https://jenkins.ceph.com/job/ceph-pull-requests-arm64/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pull-requests-arm64/config/definitions/ceph-pull-requests-arm64.yml)
- `jenkins test submodules` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-pr-submodules/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pr-submodules/config/definitions/ceph-pr-commits.yml)
- `jenkins test dashboard` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-dashboard-pull-requests/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-dashboard-pull-requests/config/definitions/ceph-dashboard-pull-requests.yml)
- `jenkins test dashboard cephadm` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-dashboard-cephadm-e2e/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-dashboard-cephadm-e2e/config/definitions/ceph-dashboard-cephadm-e2e.yml)
- `jenkins test api` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-api/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pr-api/config/definitions/ceph-pr-api.yml)
- `jenkins test docs` [ReadTheDocs](https://readthedocs.org/projects/ceph/) | [Github Workflow Definition](https://github.com/ceph/ceph/blob/main/.readthedocs.yml)
- `jenkins test ceph-volume all` [Jenkins Jobs](https://jenkins.ceph.com/view/ceph-volume%20PR/) | [Jenkins Jobs Definition](https://github.com/ceph/ceph-build/blob/main/ceph-volume-cephadm-prs/config/definitions/ceph-volume-pr.yml)
- `jenkins test windows` [Jenkins Job](https://jenkins.ceph.com/job/ceph-windows-pull-requests/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-windows-pull-requests/config/definitions/ceph-windows-pull-requests.yml)
- `jenkins test rook e2e` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-orchestrator-rook-e2e/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-rook-e2e/config/definitions/ceph-orchestrator-rook-e2e.yml)
</details>

sseshasa committed:
src/common/mclock_common: Fix output formatting of SchedulerClass

The earlier output formatting was resulting in the value and string
representation of the SchedulerClass being clubbed together for
e.g., "3client"

The formatting is now fixed to log SchedulerClass as "3 (client)".

Signed-off-by: Sridhar Seshasayee <sridhar.seshasayee@ibm.com>

sseshasa committed:
osd/scheduler/mClockScheduler: Log the size of high priority queues.

Signed-off-by: Sridhar Seshasayee <sridhar.seshasayee@ibm.com>

sseshasa committed:
osd/scheduler/mClockScheduler: Fix line alignments

Signed-off-by: Sridhar Seshasayee <sridhar.seshasayee@ibm.com>

sseshasa committed:
osd/scheduler: Classify EC subOp reads according to op priority for mClock

The change brings MSG_OSD_EC_READ into the fold of mClock scheduler. This
improves the scheduling of client and other classes of operation as they
are no longer unnecessarily preempted by the 'immediate' queue.
EC SubOps are now handled as follows:

 - EC SubOp reads generated during recovery will either go into the
   'background_recovery' or 'background_best_effort' class based on
   the recovery priority set for the op. EC SubOp reads generated due
   to client will continue to be classified as 'immediate'.

 - EC SubOp writes generated as a result of client operations will
   continue to be classified as 'immediate'.

 - EC SubOp replies are considered high priority and therefore
   continue to be classed as 'immediate'.

Fixes: https://tracker.ceph.com/issues/71655
Signed-off-by: Sridhar Seshasayee <sridhar.seshasayee@ibm.com>

sseshasa committed:
src/messages, osd: Calculate and set cost for subOpReads for mClock scheduler

Previously, sub-op reads returned a hardcoded cost of 0, bypassing
mClock's background bandwidth and tag calculation mechanisms. This
allowed backfill operations to proceed un-metered, occasionally causing
backend resource contention and driving up client tail latencies.

Cost is calculated based on whether the complete chunk/shard or a subchunk
needs to be read. The possible cases are:
1. Read the complete chunk aligned length:
   - Cost is set to the length of the chunk aligned extent size.
2. Fragmented reads:
   - Consider the subchunk length and count to calculate the cost.
   - compute_cost evaluates the exact layout of fragmented shard bytes on
     disk by summing up the active subchunk allocations exactly once
     (`fragmented_shard_bytes += k.second * subchunk_size`).
   - Linear Extent Scaling: Scale the baseline footprint cleanly by
     multiplying it against the true count of read extents (`tl.size()`),
     achieving a highly efficient O(N) time complexity.

This linear cost model is compatible with pools running with
'allow_ec_optimizations' set to true. Under the FastEC optimized
pipeline, most operations are unified and bypass fragment slicing,
meaning requests will primarily match the Case 1 chunk-aligned path.
In Case 2 where applicable, the O(N) loop ensures that cost will
scale proportionally according to the layout.

It is important to note that the amount of data to read was set to an upper
bound defined by osd_recovery_max_chunk (8 MiB) and was rounded up to the
stripe width. The reason for setting a higher than actual upper bound is that
there may be cases where the object doesn't have the xattrs yet to determine
its size. Therefore, the amount to read was ultimatly set to ~(8 MiB / k)
where k is the number of data shards. This can cause mClock to prolong
the recovery times as items stay longer in the queue. To address this, the
amount to read is set to the remaining length of the object to recover
if the object size is known. Otherwise, the amount to read is set to the
recovery chunk size as before. Therefore, in some cases, only the first
recovery read could be costly if the object context is not known.

The MOSDECSubOpRead class introduces the following:
 - cost member. This necessitates an increment to the HEAD_VERSION and
   appropriate handling within the encode and decode methods.
 - compute_cost() that is called when creating the message by
   ECCommonL::ReadPipeline::do_read_op(). This calls into ECSubRead::cost()
   that performs the actual calculations to set the cost based on the cases
   mentioned above.
 - The same sequence applies to the EC optimized path in
   ECCommon::ReadPipeline::do_read_op().

Fixes: https://tracker.ceph.com/issues/71655
Signed-off-by: Sridhar Seshasayee <sridhar.seshasayee@ibm.com>

sseshasa commented:
@athanatos Following up from our discussion, I think I understand what needs to change when determining the cost for a subOpRead. In `ECBackendL::RecoveryBackend::continue_recovery_op()`, the `amount` in the `RecoveryOp::IDLE:` case is set by calling `get_recovery_chunk_size()` (see: https://github.com/ceph/ceph/blob/5b828b6a44549db27bb6bafa11af82f832931f67/src/osd/ECBackendL.cc#L575). 

For mClock this should be changed to return the object size being recovered and rounded up to the stripe width. Otherwise, this will always be set close to the recovery chunk size (~8 MiB) which is incorrect. Do you agree with this?

athanatos commented on src/messages/MOSDECSubOpRead.h:
```diff
@@ -20,16 +20,23 @@
+  void compute_cost(CephContext *cct, const std::set<hobject_t> &hoids) {
+    cost = 0;
+    for(const auto& hoid : hoids) {
+      cost += op.cost(cct, hoid);
```
Why does compute_cost here need hoids passed in?  Doesn't the ECSubRead op already know what objects it is requesting?

athanatos commented on src/osd/ECMsgTypes.cc:
```diff
@@ -252,6 +252,22 @@ void ECSubRead::decode(bufferlist::const_iterator &bl)
+  if (cct->_conf->osd_op_queue == "mclock_scheduler") {
+    // 'to_read' (map<hobject_t, list<tuple<offset, length, flags>>>)
+    for (auto& rl : to_read[hoid]) {
+      cost += boost::get<1>(rl);
```
Don't we assign a fixed cost for each IO in addition to a variable cost based on the length?

Why pass in hoid?  Can't ECSubRead simply return the total cost of all of the contained reads?

ECSubRead also has a subchunks member related to the new optimizations, that should be included as well, I think.

sseshasa commented:
> Ah, we submit ECSubRead with the max amount and trust the other side to truncate. There are cases where it has to work that way -- the first request may not actually know the object size since it doesn't have object info xattr yet. I don't think this is worth messing with for now, just use the value actually in the ECSubRead. We can make it more accurate in a followup PR.

Sure, I will calculate the value from ECSubRead's `std::map<hobject_t, std::list<boost::tuple<uint64_t, uint64_t, uint32_t> >> to_read`  to calculate the cost. I will address your other comments as well.

sseshasa commented on src/osd/scheduler/mClockScheduler.cc:
```diff
@@ -106,11 +106,16 @@ void mClockScheduler::enqueue(OpSchedulerItem&& item)
-          << dendl;
+  dout(20) << __func__ << "sched client_count: " << scheduler.client_count()
+           << " sched queue size: " << scheduler.request_count()
+           << dendl;
```
Sure, I have made a note of this.

sseshasa commented on src/osd/scheduler/mClockScheduler.cc:
```diff
@@ -106,11 +106,16 @@ void mClockScheduler::enqueue(OpSchedulerItem&& item)
-	  << " high_priority_queue: " << high_priority.size()
-          << " sched: " << scheduler.request_count() << " ]"
-          << dendl;
+  dout(20) << __func__ << "sched client_count: " << scheduler.client_count()
```
Done.

sseshasa commented on src/osd/scheduler/OpSchedulerItem.h:
```diff
@@ -232,15 +232,52 @@ class PGOpItem : public PGOpQueueable {
+     *    to be classed as 'immediate'.
+     *
+     *  Note: The 'cost' for subOp read and write operation is set to 1 as
+     *  before so that they don't spend time in the scheduler queue.
```
Yes, I will update this in the upcoming pushes.

sseshasa replied:
Updated the comment.

sseshasa commented:
jenkins test api

sseshasa commented:
jenkins test make check arm64

sseshasa commented:
jenkins test windows

sseshasa commented:
jenkins test windows

sseshasa commented:
jenkins test windows

sseshasa commented:
> One solution is to read the size from the object info xattrs and use it if valid or set the cost to the existing upper bound otherwise. I am not clear on how often we may not get the xattrs from the object at this point during recovery. What do you think?

@athanatos I implemented the above solution. Please take a look at the latest push. I ran multiple tests with this change and did not observe any issue.

sseshasa commented:
jenkins test make check

sseshasa commented:
jenkins test api

sseshasa commented:
jenkins test windows

sseshasa commented:
jenkins test make check

sseshasa commented:
jenkins test windows

rzarzynski commented:
jenkins test windows

markhpc commented:
@sseshasa Any news on this one?

sseshasa committed:
mclock_common, mClockScheduler: Add perf counters for scheduler ops

Add perf counters to show the status pertaining to the number of ops,
dynamic queue lengths, queue latency and bytes read for the following
ops handled in the high queues and in the scheduler queues:
 - peering
 - client
 - ec reads/writes
 - ec recovery reads

Additional counters can be added in the future based on the requirement.

Signed-off-by: Sridhar Seshasayee <sridhar.seshasayee@ibm.com>

sseshasa commented:
Update:
Tests are in progress. This PR can be considered as WIP until tests are completed including teuthology.

sseshasa commented:
jenkins test make check

markhpc commented:
@sseshasa Heya, any news on progress here?  If we think this is going to continue taking a while, perhaps we should add a note to the docs about switching back to WPQ?

sseshasa commented:
> @sseshasa Heya, any news on progress here? If we think this is going to continue taking a while, perhaps we should add a note to the docs about switching back to WPQ?

Hi @markhpc , The good news is that the teuthology tests look good with the latest change. I am currently testing the scenario using a couple of EC profiles to validate no slow-ops show up and see if there's an improvement in client tail latency. I hope to get this to closure in the next week or so pending another of reviews.

sseshasa commented:
The latest push (May 21, 2026) adds a bunch of perf counters to help track the queue length and latency of different ops handled within mClock queue and the high_queues. This would be helpful in tracking down issues observed if any during disruptive operations with EC based clusters.

**Testing Progress Notes**

1. I am testing the changes on ec2+1 and ec4+2 profiles on a NVMe based cluster. (80% completed).
2. Unfortunately, HDDs are not yet available on the machines in the new lab and so I have to engage the testing team for this.
3. But specific data from the current test results would be useful indicators (for e.g. client latency differences between runs with and without fix) on how the fix is helping. 
4. Once the test runs complete (early next week), I will update the results and comparison.
5. Teuthology test runs look good (will update the results as well here).
6. Once the reviews are complete, the plan is to test the fix on a larger HDD based cluster before merging this upstream and downstream.
7. This fix can be tested by upstream users if possible and it would be great to hear feedback.

sseshasa committed:
mclock_common: adjust mClock profile parameters to prevent backfill starvation

Adjust the 'background_best_effort' queue parameters across the
three standard mClock profiles (high_client_ops, balanced, and
high_recovery_ops) to ensure best effort ops are not starved.

Previously, the 'background_best_effort' queue carried a default allocation
of 0% (MIN) reservation and a weight of 1 under these profiles. When
concurrent client traffic is dense, the zero-reservation for example completely
starves backfill sub-ops (MSG_OSD_EC_READ) on pools with
'allow_ec_optimizations' set to false. This starvation forces the Primary OSD
to hold internal BlueStore transactions and PG object locks for extended
windows, causing severe client median (50th) latency inflation.

To prevent background starvation and resolve the effects of the primary lock
retention, the profile configurations are tuned as follows:

The following profile changes forces low-cost sub-ops to clear out of peer
queues rapidly to drop  primary locks, which helps improve the client
completion latency and tail latency (95th, 99th and 99.5th) percentile.

1. high_client_ops profile:
   - Grant 'background_best_effort' a safe 5% minimum reservation.
   - Scale the queue weight to 4.

2. balanced profile:
   - Grant 'background_best_effort' a 5% minimum reservation.
   - Set the queue weight to 2.

3. high_recovery_ops profile:
   - Grant 'background_best_effort' a 5% minimum reservation.
   - Set the queue weight to 2.

4. Modify the mClock config reference documentation to reflect the tuning
   changes to the best-effort QoS parameters across the profiles.

Note on Proportional Scaling Compatibility:
Configuring these changes shifts total reservations to 105% (e.g., 50%
client + 50% recovery + 5% best-effort under the Balanced profile). Under
heavy concurrent saturation, mClock's internal controls resolves this
gracefully via proportional down-scaling, preserving the underlying
device bandwidth limits for different classes of clients. For example instead
of the client being allocated 50% bandwidth, a slightly lower reservation is
allocated while shifting the remaining bandwidth to the best-effort queue.
This minor scaling shift is virtually unnoticeable to the client application,
but it prevents the internal queue deadlocks.

Signed-off-by: Sridhar Seshasayee <sridhar.seshasayee@ibm.com>

sseshasa commented:
# Optimizing Erasure Coded Sub-Op Read Scheduling Under mClock - Test Report

## 1. Summary

During the execution of concurrent client workloads and background backfill operations on Erasure Coded (EC) pools, slow Ops along with client tail latency degradation was observed. Slow ops were reported because client ops were preempted by a surge of EC subOp reads in the higher priority 'immediate' queue due to marking an OSD 'OUT' for draining. This report evaluates a two-fold solution combining a Cost Fix with mClock Profile Allocation Adjustments on an NVMe based Ceph cluster.

Across all three standard profiles (_high_client_ops_, _balanced_, and _high_recovery_ops_), the combined solution successfully eliminated Primary OSD lock contention, stabilized background data movement pacing, and compressed client tail latencies below cluster baselines.

## 2. Test Environment
Nodes: 1 Nos,  Architecture:  x86_64, 192 Cores, 500 GiB Memory
OSDs: 7 OSDs, Created on 1:1 Mapping on NVMe devices (Model: SAMSUNG MZQL27T6HBLA-00A07)
EC Profile: 4+2, plugin=isa technique=reed_sol_van
IO Tool: Fio, QD: 64
Tested with ~130K Objects with object sizes in range 1 KiB - 16 MiB.

**NOTE: The Fio client  was configured to run at 80% of the maximum measured benchmark to simulate a realistic workload.**

## 3. Problem Statement & Root Cause

The first part of the solution simply redirected the EC subOp reads to the appropriate '_background-recovery_' or the '_background-best-effort_' queue based on the recovery priority. But on EC pools configured with legacy chunk layout handling (allow_ec_optimizations: false), moving EC sub-op reads (`MSG_OSD_EC_READ`) from the unthrottled immediate queue to mClock background queues exposed a critical structural bottleneck:

1. The background_best_effort queue (where backfills are enqueued) carried a 0% (MIN) reservation. Dense client traffic completely starved this queue, indefinitely extending lock retention windows.

2. While subOp reads sat starved in replica OSD mClock queues, the Primary OSD remained blocked waiting for replies, continuously holding internal BlueStore transactions, Placement Group (PG), and Object locks. Incoming client operations bypassed the mClock queue cleanly but stalled at the object layer waiting for backend transaction locks to drop, causing latency inflation.

The results of the tests with the cost fix but unmodified profile allocations shared below.

### 3.1. Test Results with Cost Fix Alone & Comparison With Baseline (No Fix)
With a 0% default reservation caused backfills to be starved whenever client traffic peaked. This starvation forced the Primary OSD to hold internal object and transaction locks open for too long. Client requests passed through mClock instantly, but immediately encountered those backend lock blockades, which is why the 50th percentile spiked to 5.210 ms. This test proved that fixing the cost alone was not enough; a nominal reservation floor was absolutely mandatory to cleanly flush the locks.

**Client Performance Matrix (Cost Fix alone)**

| Metric | Baseline (No Fix) | Cost Fix Alone | Performance Status vs. Baseline |
| :--- | :---: | :--- | :--- |
| **Throughput (IOPS)** | 8647.94 | 8628.89 | Minor Workload Slump |
| **Avg clat (msec)** | 5.455 | 6.206 | Degraded (+25.75% Latency Inflation) |
| **50th pctl (msec)** | 4.492 | 5.210 | Degraded (+29.05%) |
| **95th pctl (msec)** | 11.469| 13.041 | Degraded (+26.10%) |
| **99.5th pctl (msec)** | 21.050| 23.199 | Degraded (+17.52%) |


## 4. Implemented Solutions

The following solution which is two-fold was implemented:

1. Set cost for EC recovery reads: The cost function was written to compute the exact physical byte footprint of full or fragmented chunks linearly. It maps the memory allocations exactly once and scales it directly against the read extent count.

2. Profile allocation tuning: To establish an unstarvable queue that allows background subOps to clear quickly and drop primary locks, the background_best_effort profile settings were updated globally:
    * high_client_ops: res = 5% [was 0%], wgt = 4 [was 1], lim = 70%
    * balanced: res = 5% [was 0%], wgt = 2 [was 1], lim = 90%
    * high_recovery_ops: res = 5% [was 0%], wgt = 2 [was 1], lim = MAX

## 5. Performance Results
### 5.1. Profile 1: balanced Performance

#### Comparison Charts

<img width="948" height="352" alt="balanced_profile_Avg_Client_Latency_Comparison_During_Misplaced_Recovery_NVMe" src="https://github.com/user-attachments/assets/b0faaefc-1a44-412f-8290-8f59d0d2b3e0" />
<img width="789" height="352" alt="balanced_profile_Avg_Misplaced_Objects_Recovered_Time_Taken_NVMe" src="https://github.com/user-attachments/assets/362defdc-3a31-4ee4-8d75-0cabc2870cd5" />
<img width="586" height="331" alt="balanced_profile_Recovery_Throughput_Comparison" src="https://github.com/user-attachments/assets/a8e19f8e-53da-4cd6-9c76-f34e89f3f49e" />


#### Client Performance Matrix (Comparison)

| Metric | Baseline (No Fix) | Combined Solution (Cost Fix + Profile Mod) | Delta (Combined vs Baseline) |
| :--- | :---: | :---: | :--- |
| **Throughput (IOPS)** | 8647.94 | **8649.02** | +0.01% [Stable] |
| **Avg clat (msec)** | 5.455 | **4.937** | -9.49% [Improvement] |
| **50th pctl (msec)** | 4.492 | **4.047** | -9.91% [Improvement] |
| **95th pctl (msec)** | 11.469 | **10.715** | -6.57% [Improvement] |
| **99.5th pctl (msec)** | 21.050 | **21.313** | +1.25% [Stable] |

**Observed Improvements:**

* Client completion latency fell by 9.49% and median latency fell by 9.91%, proving that the 5% reservation successfully flushed subOps to release primary locks and optimize client Ops.
* Backfill speed: The erratic baseline bursts (stdev = 31.46) was controlled down to 13.24 (a 57.92% stabilization), preventing device queues from thrashing.
* Elimination of Object Bloat: Mean objects recovered dropped by 15.96% because unblocking backfill ops eliminated ongoing client/backfill state re-mappings.

### 5.2. Profile 2: high_client_ops Performance

#### Comparison Charts

<img width="948" height="352" alt="high_client_ops_profile_Avg_Client_Latency_Comparison_During_Misplaced_Recovery_Test_NVMe" src="https://github.com/user-attachments/assets/df94678a-7e35-4e9f-994a-0150474ac0de" />
<img width="789" height="352" alt="high_client_ops_profile_Avg_Misplaced_Objects_Recovered_Time_Taken_NVMe" src="https://github.com/user-attachments/assets/68f801e0-9b64-4e1d-b093-e088bad310e9" />
<img width="598" height="331" alt="high_client_ops_profile_Recovery_Throughput_Comparison" src="https://github.com/user-attachments/assets/f6c64f7a-5e6e-43b1-bd02-15f6cd5f38bc" />

#### Client Performance Matrix (Comparison)

| Metric | Baseline (No Fix) | Combined Run (Cost Fix + Profile Mod) | Delta (Combined vs Baseline) |
| :--- | :---: | :---: | :---: |
| **Throughput (IOPS)** | 8649.75 | **8647.13** | -0.03% [Stable] |
| **Avg clat (msec)** | 4.935 | **4.593** | -7.01% [Improvement] |
| **50th pctl (msec)** | 4.037 | **3.854** | -4.53% [Improvement] |
| **95th pctl (msec)** | 10.342 | **9.287** | -10.20% [Improvement] |
| **99.5th pctl (msec)** | 19.740 |**18.494** | -6.31% [Improvement] |


**Observed Improvements:**

* Latency Compression: The results show a reduction in average clat and pushed median (50th) latency to an all-time low of 3.854 ms showing that the combined fix successfully clears out the lock delays.
* Tail Stabilization: The two-fold fix preserves strong tail compression at the 95th and 99.5th percentiles while stabilizing backfill delivery down to a standard deviation of 15.71.

### 5.3. Profile 3: high_recovery_ops Performance

#### Comparison Charts

<img width="948" height="352" alt="high_recovery_ops_profile_Avg_Client_Latency_Comparison_During_Misplaced_Recovery_Test_NVMe" src="https://github.com/user-attachments/assets/3a990b89-8165-4fae-bff7-2bc26ba88c98" />
<img width="797" height="352" alt="high_recovery_ops_profile_Avg_Misplaced_Objects_Recovered_Time_Taken_NVMe" src="https://github.com/user-attachments/assets/2b012205-ce5a-4d13-b836-a77f54cfef85" />
<img width="598" height="331" alt="high_recovery_ops_profile_Recovery_Throughput_Comparison " src="https://github.com/user-attachments/assets/c3e5fe67-c135-418b-beff-8705b704f3a5" />


#### Client Performance Matrix (Comparison)

| Metric | Baseline (No Fix) | Combined Run (Cost Fix + Profile Mod) | Delta (Combined vs Baseline) |
| :--- | :---: | :---: | :---: |
| **Throughput (IOPS)** | 8638.32 | **8648.95** | +0.12% [Stable] |
| **Avg clat (msec)** | 4.802 | **4.904** | +2.12% [Stable & Within Device Margin] |
| **50th pctl (msec)** | 4.027 | **4.165** | +3.42% [Stable & Within Device Margin] |
| **95th pctl (msec)** | 9.863 | **9.896** | +0.33% [Stable] |
| **99.5th pctl (msec)** | 18.743 | **18.691** | -0.28% [Improvement] |


**Observed Improvements:**

* Preserve Client Performace: The 30% default client reservation pre-throttled incoming writes, completely insulating client median (4.165 ms) and tail (18.691 ms) latencies from background activity. There was a negligible increase in the clat and tail latency with the introduction of a weight of 2 and 5% reservation. The performance of this profile is largely preserved.

## 6. Fix Compatibility & Hardware Adaptability

This combined solution is compatible across both Flash (SSD/NVMe) and Hard Drive (HDD) environments:

* On SSD/NVMe: The two-fold solution maximizes parallel flash channel utilization, translating directly into lower average client completion metrics as observed in the performance results.


* On HDDs: The combined solution feeds mClock accurate physical byte footprints. This keeps mClock tag intervals tight and accurate. Paired with the dedicated, stable 5% reservation floor, mClock can cleanly interleave client blocks and backfill fragments in predictable time slices. This should help preserve native drive pacing, prevent thrashing, and ensure stable throughput. Tests on HDD environment is the next step which will help validate the solution further.

## 7. Conclusion & Next Steps

As seen in the results, the combined solution works well in terms of isolating client and backfill traffic and prevent one class of operation from indirectly blocking another within the PG & BlueStore layers.

Although not tested in the current effort, the modified reservation and weight parameters of the background-best-effort class should help other types of best-effort operations like scrub and snaptrim to also perform better. If these operations abuse the slightly expanded weight/reservation configuration, there are existing safe guards to limit them from running away unchecked.

Therefore, this fix acts as a good starting point to address some outstanding issues with mClock's background operation scheduling with more fixes coming up.

sseshasa committed:
doc/rados/configuration: Remove wpq recommendation warning for EC clusters

Remove the warning that recommends using wpq scheduler as a fallback for EC
clusters. This issue is addressed by considering EC recovery reads as
background, assigning an accurate cost for those reads and tuning the QoS
parameters associated with best-effort class of operations.

Signed-off-by: Sridhar Seshasayee <sridhar.seshasayee@ibm.com>

sseshasa commented:
### Teuthology Test Result

https://pulpito.ceph.com/sseshasa-2026-05-26_18:29:48-rados-wip-sseshasa-testing-2026-05-26-1779803771-distro-default-trial/
270/295 Passed, 25 Failed

### All Failures are Known and Unrelated

1. ['238505'] - https://tracker.ceph.com/issues/76298 - rados/thrash-erasure-code failure - not all PGs are active or peered (RADOS)
2. ['238488', '238356'] - https://tracker.ceph.com/issues/76297 - test_iscsi_container - cephadm/test_iscsi_setup.sh fails on sudo ls /dev/disk/by-path - (Orchestrator)
3. ['238486'] - https://tracker.ceph.com/issues/75488 - qa/standalone/osd/osd-recovery-prio.sh: TEST_recovery_pool_prio: The primary PG for testX didn't become the in progress item [on remote] - (RADOS)
4. ['238478', '238413', '238284'] - https://tracker.ceph.com/issues/76497 - rados/singleton-bluestore/cephtool - evicting unresponsive client warnings - (RADOS)
5. ['238446'] - https://tracker.ceph.com/issues/65719 - debian-17.2.6 jammy repository does not have a Release file - (RADOS)
6. ['238445'] - https://tracker.ceph.com/issues/76473 - rados/valgrind-leaks: expected valgrind issues and found none - (RADOS)
7. ['238434] - https://tracker.ceph.com/issues/76981 - Rocky10 - test_cephadm.sh failure cause by Cannot register NFS V3 on TCP - (Orchestrator)
8. ['238427'] - https://tracker.ceph.com/issues/74827 - rados/standalone/mon/misc.sh - TEST_mon_features return 1 (RADOS)
9. ['238399', '238251'] - https://tracker.ceph.com/issues/75414 - upgrade: pool full warning variation - (RADOS)
10. ['238395', '238244'] - https://tracker.ceph.com/issues/70669 - ERROR: test_list_enabled_module: cephfs resource temporarily unavailable - (Dashboard)
11. ['238346'] - https://tracker.ceph.com/issues/72888 - rados/singleton-bluestore cluster [ERR] overall HEALTH_ERR 1 auth entities have invalid capabilities - (Ceph)
12. ['238338'] - https://tracker.ceph.com/issues/76470 - rados/test_crash.sh: Induced crashes are not properly posted - (RADOS)
13. ['238323'] - https://tracker.ceph.com/issues/74794 - rados/thrash - The WatchDog barked due to Thrasher ceph.osd.9 threw exception wait_for_recovery: failed before timeout expired - Many PGs stuck not active+clean - (RADOS)
14. ['238300', '238243', '238242', '238294'] - https://tracker.ceph.com/issues/74708 - Failed to download metadata for repo 'epel' causes 'sudo yum install -y kernel' to fail on trial129
15. ['238291', '238271'] - https://tracker.ceph.com/issues/71344 - Monitor Reconnection Failure and MON_NETSPLIT due to Stale Global Sequence Check During Incoming ConnectionMonitor Reconnection Failure and MON_NETSPLIT due to Stale Global Sequence Check During Incoming Connection - (RADOS)
16. ['238245'] - https://tracker.ceph.com/issues/74074 - ceph-dencoder - failed with 20.3 corpus repo - (RADOS)

rzarzynski commented on src/common/mclock_common.cc:
```diff
@@ -34,23 +34,29 @@ namespace dmc = crimson::dmclock;
-    return lhs;
+    lhs << "unknown";
   }
+  return lhs << ")";
```
ACK.

rzarzynski commented on src/osd/scheduler/mClockScheduler.cc:
```diff
@@ -99,11 +99,19 @@ void mClockScheduler::enqueue(OpSchedulerItem&& item)
+  for (const auto& [prio, queue] : high_priority) {
+    *_dout << ", priority " << fmt_prio(prio) << ": " << queue.size();
+  }
+  *_dout << dendl;
```
ACK.

rzarzynski commented on src/osd/scheduler/OpSchedulerItem.h:
```diff
@@ -237,12 +237,47 @@ class PGOpItem : public PGOpQueueable {
+     *
+     *  Note: The 'cost' for EC subOp read operation is set according to
+     *  the amount of data to be read/written.
+     */
```
ACK.

rzarzynski commented on src/messages/MOSDECSubOpRead.h:
```diff
@@ -58,6 +68,9 @@ class MOSDECSubOpRead : public MOSDFastDispatchOp {
       min_epoch = map_epoch;
     }
+    if (header.version >= 4) {
+      decode(cost, p);
```
Please note the `cost` member is default-initialized to `0`:

```cpp
  uint64_t cost = 0;
```

so we should be fine also with `version < 4`. ACK.

rzarzynski commented on src/osd/ECBackendL.cc:
```diff
@@ -573,7 +573,20 @@ void ECBackendL::RecoveryBackend::continue_recovery_op(
+      if (op.obc) {
+        uint64_t remaining = op.obc->obs.oi.size - from;
+        if (remaining < amount) {
+          amount = remaining;
```
nit: `std::min(remaining, amount)`.

sseshasa replied:
Done. Made the change.

rzarzynski commented on src/osd/ECCommonL.cc:
```diff
@@ -489,6 +492,7 @@ void ECCommonL::ReadPipeline::do_read_op(ReadOp &op)
       msg->trace.init("ec sub read", nullptr, &op.trace);
       msg->trace.keyval("shard", i->first.shard.id);
     }
+    msg->compute_cost(cct, subchunk_info);
     m.push_back(std::make_pair(i->first.osd, msg));
```
My understanding is an OSD can make use of the cost only if its peer has calculated it (which means the peer: 1) runs recent enough version and 2) operates with `cct->_conf->osd_op_queue == "mclock_scheduler"`).

sseshasa replied:
That's correct.

rzarzynski commented on src/osd/ECMsgTypes.cc:
```diff
@@ -266,6 +268,55 @@ void ECSubRead::decode(bufferlist::const_iterator &bl)
+uint64_t ECSubRead::cost(CephContext *cct, std::pair<int, int>& subchunk_info)
+{
+  uint64_t total_cost = 0;
+  if (cct->_conf->osd_op_queue != "mclock_scheduler") {
+    return total_cost; // Legacy behavior for WPQ scheduler
+  }
```
IIRC this part conveys the implicit assumption that the scheduler is the same between the OSD that generates and messages and the one that receives it. I think it's roughly fine and my comment is an example of nit picking but would be good to at least comment it and be explicit.

sseshasa replied:
Sure, I have added a comment here indicating how cost is set in case there's a difference in the scheduler type between primary and replica shards.

sseshasa commented:
I have merged the PR to main based on my teuthology testing (see results [here](https://github.com/ceph/ceph/pull/64618#issuecomment-4632595930)). While this bakes in main, the plan is test this on HDD based clusters and then backport it to the prior branches.

Results from testing on SSD based EC clusters is discussed above [here](https://github.com/ceph/ceph/pull/64618#issuecomment-4544159885).
